src/main/scala/nl/knaw/dans/easy/multideposit/Action.scala

```diff
@@ -85,101 +92,92 @@ trait Action extends DebugEnhancedLogging {
               footer = "Due to these errors in the preconditions, nothing was done."),
             cause = e))
       }
-      _ <- innerExecute().recoverWith {
-        case NonFatal(e) =>
-          List(Failure(e), innerRollback())
-            .collectResults
-            .recoverWith {
-              case e: CompositeException => Failure(ActionRunFailedException(
-                report = generateReport(
-                  header = "Failures during the execute phase:",
-                  throwable = e,
-                  footer = "The actions that were already performed, were rolled back."),
-                cause = e))
-            }
+      t <- innerExecute(a).recoverWith {
+        case e1@CompositeException(es) =>
+          rollback() match {
+            case Success(_) => reportFailure(e1)
+            case Failure(CompositeException(es2)) => reportFailure(CompositeException(es ++ es2))
+            case Failure(e2) => reportFailure(CompositeException(es ++ List(e2)))
+          }
+        case NonFatal(e1) =>
+          rollback() match {
+            case Success(_) => reportFailure(e1)
+            case Failure(CompositeException(es2)) => reportFailure(CompositeException(List(e1) ++ es2))
+            case Failure(e2) => reportFailure(CompositeException(List(e1, e2)))
+          }
       }
-    } yield ()
+    } yield t
   }
 
   private def generateReport(header: String = "", throwable: Throwable, footer: String = ""): String = {
-    val report = throwable match {
-      case ActionException(row, msg, _) => s" - row $row: $msg"
-      case CompositeException(throwables) =>
-        throwables.toSeq
-          .collect {
-            case ActionException(row, msg, _) => s" - row $row: $msg"
-            case NonFatal(e) => s" - unexpected error: ${e.getMessage}"
-          }
-          .mkString("\n")
-      case e => s" - unexpected error: ${e.getMessage}"
+
+    @tailrec
+    def report(es: List[Throwable], rpt: List[String] = Nil): List[String] = {
+      es match {
+        case Nil => rpt
+        case ActionException(row, msg, _) :: xs => report(xs, s" - row $row: $msg" :: rpt)
+        case CompositeException(ths) :: xs => report(ths.toList ::: xs, rpt)
+        case NonFatal(ex) :: xs => report(xs, s" - unexpected error: ${ex.getMessage}" :: rpt)
+      }
     }
 
-    header.toOption.map(_ + "\n").getOrElse("") + report + footer.toOption.map("\n" + _).getOrElse("")
+    header.toOption.fold("")(_ + "\n") +
+      report(List(throwable)).reverse.mkString("\n") +
+      footer.toOption.fold("")("\n" + _)
   }
 
   /**
-   * Compose this action with `other` into a new action such that it runs both actions in the correct order.
-   * [[checkPreconditions]] and [[execute]] are run in order, [[rollback]] is run in reversed order.
+   * Sequentially composes two `Action`s by running their `execute` methods one after the other.
+   * In this, the input of `other` is the output of this `Action`.
```

Comment:
Please more details about the execution order:
* preconditions of this
* preconditions of other
* execute of this, on failure rollback of this
* execute of other, on failure rollback of other, then rollback of this